components/org.wso2.carbon.appmgt.impl/src/main/java/org/wso2/carbon/appmgt/impl/token/AbstractJWTGenerator.java

```diff
@@ -0,0 +1,524 @@
+        try {
+            Key privateKey = getPrivateKey(endUserName, tenantId);
+            if (privateKey == null) {
+                throw new AppManagementException("Private key is null for tenant " + tenantId);
+            }
+            /* Initialize signature with private key and algorithm */
+            Signature signature = Signature.getInstance(signatureAlgorithm);
+            signature.initSign((PrivateKey) privateKey);
+
+            /* Update signature with data to be signed */
+            byte[] dataInBytes = assertion.getBytes(StandardCharsets.UTF_8);
+            signature.update(dataInBytes);
+
+            /* Sign the assertion and return the signature */
+            byte[] signedInfo = signature.sign();
+            return signedInfo;
+        } catch (NoSuchAlgorithmException e) {
+            String error = "Signature algorithm " + signatureAlgorithm + "not found.";
+            log.error(error, e);
+            throw new AppManagementException(error, e);
+        } catch (InvalidKeyException e) {
+            String error = "Invalid private key provided for the signature for tenant " +  tenantId;
+            log.error(error, e);
+            throw new AppManagementException(error, e);
+        } catch (SignatureException e) {
+            String error = "Error in signature algorithm " + signatureAlgorithm;
+            log.error(error, e);
+            throw new AppManagementException(error, e);
+        } catch (AppManagementException e) {
+            String error = "Error in obtaining tenant's" + tenantId + "private key";
+            log.error(error, e);
+            throw new AppManagementException(error, e);
+        }
+    }
+
+    /**
+     * Helper method to get private key for specific tenant.
+     *
+     * @param endUserName
+     * @param tenantId
+     * @return private key
+     * @throws org.wso2.carbon.appmgt.api.AppManagementException
+     */
+    private Key getPrivateKey(String endUserName, int tenantId) throws AppManagementException {
+
+        String tenantDomain = MultitenantUtils.getTenantDomain(endUserName);
+        try {
+            Key privateKey = privateKeys.get(tenantId);
+            if (privateKey == null) {
+                KeyStoreManager tenantKSM = getKeyStoreManager(tenantId);
+
+                if (!tenantDomain.equals(MultitenantConstants.SUPER_TENANT_DOMAIN_NAME)) {
+                    /* Derive key store name */
+                    String keyStoreName = tenantDomain.trim().replace(".", "-");
+                    String jksName = keyStoreName + ".jks";
+                    /* Obtain private key */
+                    privateKey = tenantKSM.getPrivateKey(jksName, tenantDomain);
+                } else {
+                    try {
+                        privateKey = tenantKSM.getDefaultPrivateKey();
+                    } catch (Exception e) {
+                        String error = "Error while obtaining private key for super tenant";
```

Comment:
Generic exception is thrown by org.wso2.carbon.core.util.KeyStoreManager's getDefaultPrivateKey() method. We do not have any other option in here.
https://github.com/wso2/carbon-kernel/blob/4.4.x/core/org.wso2.carbon.core/src/main/java/org/wso2/carbon/core/util/KeyStoreManager.java#L388